### PR TITLE
Add the ability to inline all kernels exhaustively

### PR DESCRIPTION
**Pull request by alan-baker**
* Gated by option
* new pass exhaustively inlines all kernels
* added a test

### CODE REVIEW

File: lib/InlineEntryPointsPass.cpp [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  // Clean up dead functions. This done here to avoid ordering requirements on
+  // inlining.
+  std::vector<Function *> to_delete;
+  for (auto &F : M) {
+    if (F.isDeclaration() || F.getCallingConv() == CallingConv::SPIR_KERNEL)
+      continue;
+
+    if (F.user_empty())
```
Comment (dneto0) [inline lib/InlineEntryPointsPass.cpp]: I would have put this in a convergence loop.  We might visit a leaf function before one of its callees.  In which case the leaf sticks around, incorrectly.

Comment (alan-baker) [reply lib/InlineEntryPointsPass.cpp]: This example works and is roughly the same:
```
int func_3(local int *in, int n) { return in[n]; }
int func_2(local int *in, int n) { return func_3(in, n); }
int func_1(local int *in, int n) { return func_2(in, n); }
kernel void kernel_1(local int *in, global int *out, int n) {
  out[n] = func_1(in, n);
}
kernel void kernel_2(local int *in, global int *out, int n) {
  out[n] = func_1(in, n);
}

int foo(local int *in, int n) { return func_3(in, n); }
```

I can't explain the difference. For your example LLVM reports that both inner() and outer() have 2 uses before inlining and 1 afterwards. For my example all the uses reported make sense and the cleanup removes all function except the kernels.

Comment (dneto0) [reply lib/InlineEntryPointsPass.cpp]: I can't explain those stray uses.  Weird.

**Comment by dneto0** [comment]
Rebased, squashed, and pushed into master as f8661418a0ae

**Comment by dneto0** [comment]
Whoops.  I had the wrong PR.

**Comment by dneto0** [comment]
Rebased and pushed into master as af289ab156b

**Pull request closed without merging**